src/actions/move.ts

```diff
@@ -168,4 +184,30 @@ export class MoveActions {
       ContextMenuRegistry.registry.unregister(menuItem.id);
     }
   }
+
+  /**
+   * Get the source block for the cursor location, or undefined if no
+   * source block can be found.
+   * If the cursor is on a shadow block, walks up the tree until it finds
+   * a non-shadow block to drag.
+   *
+   * @param workspace The workspace to inspect for a cursor.
+   * @returns The source block, or undefined if no appropriate block
+   *     could be found.
+   */
+  getCurrentBlock(workspace: WorkspaceSvg): BlockSvg | undefined {
+    const curNode = workspace?.getCursor()?.getCurNode();
+    let block = curNode?.getSourceBlock();
+    if (!block) return undefined;
+    while (block && block.isShadow()) {
```

Comment:
I'm assuming the truthy check for `block` here is because TypeScript doesn't realize that it's impossible for it to ever be null or undefined?

If it's null inference is actually good enough to detect that, you should be able to drop that part of the condition.